usr/src/uts/common/fs/zfs/zap.c

```diff
@@ -505,6 +505,16 @@ zap_get_leaf_byblk(zap_t *zap, uint64_t blkid, dmu_tx_t *tx, krw_t lt,
 
 	ASSERT(RW_LOCK_HELD(&zap->zap_rwlock));
 
+	/*
+	 * If system crashed just after dmu_free_long_range in zfs_rmnode, we
+	 * would be left with an empty xattr dir in delete queue. blkid=0
+	 * would be passed in when doing zfs_purgedir. If that's the case we
+	 * should just return immediately. The underlying objects should
+	 * already be freed, so this should be perfectly fine.
+	 */
+	if (blkid == 0)
```

Comment:
I didn't address the root cause... ZoL did.  I'm just upstreaming for a customer of mine!  :)

I have what I thought was a one-time hack, but it may be more effective.  It turns some ASSERT()s into error-returns in zap_deref_leaf():

```diff
--- a/usr/src/uts/common/fs/zfs/zap.c
+++ b/usr/src/uts/common/fs/zfs/zap.c
@@ -572,9 +572,17 @@ zap_deref_leaf(zap_t *zap, uint64_t h, dmu_tx_t *tx, krw_t lt, zap_leaf_t **lp)
        uint64_t idx, blk;
        int err;
 
-       ASSERT(zap->zap_dbuf == NULL ||
+/* Change these asserts() into error throws. */
+
+       /* ASSERT(zap->zap_dbuf == NULL ||
            zap_f_phys(zap) == zap->zap_dbuf->db_data);
-       ASSERT3U(zap_f_phys(zap)->zap_magic, ==, ZAP_MAGIC);
+       ASSERT3U(zap_f_phys(zap)->zap_magic, ==, ZAP_MAGIC); */
+
+       if (zap->zap_dbuf != NULL && zap_f_phys(zap) != zap->zap_dbuf->db_data)
+               return (EIO);
+       if (zap_f_phys(zap)->zap_magic != ZAP_MAGIC)
+               return (EIO);
+
        idx = ZAP_HASH_IDX(h, zap_f_phys(zap)->zap_ptrtbl.zt_shift);
        err = zap_idx_to_blk(zap, idx, &blk);
        if (err != 0)
```

Digging in further, though, can you confirm/deny that zap_lockdir() occurs BEFORE the leaf dereference?  If so, your zap_lockdir() idea is clearly better.